algorithms/src/traits/snark.rs

```diff
@@ -26,27 +28,48 @@ pub trait Prepare<T> {
 }
 
 pub trait SNARK {
-    type AllocatedCircuit;
-    type Circuit;
+    type ScalarField: Clone + PrimeField;
+    type BaseField: Clone + PrimeField;
+
     type PreparedVerifyingKey: Clone;
     type Proof: Clone + Debug + ToBytes + FromBytes;
     type ProvingKey: Clone + ToBytes + FromBytes;
     type VerifierInput: ?Sized;
+
+    // We can specify their defaults to `()` when `associated_type_defaults` feature becomes stable in Rust
+    type UniversalSetupConfig: Clone;
+    type UniversalSetupParameters: Clone;
+
     type VerifyingKey: Clone
         + ToBytes
         + FromBytes
         + Prepare<Self::PreparedVerifyingKey>
         + From<Self::PreparedVerifyingKey>
-        + From<Self::ProvingKey>;
+        + From<Self::ProvingKey>
+        + ToConstraintField<Self::BaseField>;
 
-    fn setup<R: Rng>(
-        circuit: &Self::Circuit,
+    fn circuit_specific_setup<C: ConstraintSynthesizer<Self::ScalarField>, R: Rng>(
+        circuit: &C,
         rng: &mut R,
     ) -> Result<(Self::ProvingKey, Self::VerifyingKey), SNARKError>;
 
-    fn prove<R: Rng>(
+    fn universal_setup<R: Rng>(
```

Comment:
Explain: By default, `universal_setup` and `index` are not implemented, as in the case of Groth16 and GM17. This is reserved for Marlin, which has a slightly different API.